mm2src/mm2_metrics/src/native.rs

```diff
@@ -0,0 +1,512 @@
+        }
+    }};
+}
+
+/// Market Maker Metrics, used as inner to get metrics data and exporting
+#[derive(Default, Clone)]
+pub struct Metrics {
+    pub recorder: Arc<MmRecorder>,
+}
+
+impl Metrics {
+    /// Collect the metrics in Prometheus format.
+    pub fn collect_prometheus_format(&self) -> Result<String, String> {
+        let prometheus = PrometheusBuilder::new()
+            .set_quantiles(QUANTILES)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        Ok(prometheus.0.handle().render())
+    }
+}
+
+impl MetricsOps for Metrics {
+    fn init(&self) -> Result<(), String> {
+        Metrics::default();
+        Ok(())
+    }
+
+    fn init_with_dashboard(&self, log_state: LogWeak, interval: f64) -> Result<(), String> {
+        let recorder = self.recorder.clone();
+        let runner = TagObserver::log_tag_metrics(log_state, recorder, interval);
+        spawn(runner);
+
+        Ok(())
+    }
+
+    /// Collect prepared metrics json from the recorder
+    fn collect_json(&self) -> Result<Value, String> {
+        serde_json::to_value(self.recorder.prepare_json()).map_err(|err| ERRL!("{}", err))
+    }
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct TagMetric {
+    pub tags: Vec<Tag>,
+    pub message: String,
+}
+
+#[derive(PartialEq, PartialOrd)]
+pub enum PreparedMetric {
+    Unsigned(u64),
+    Float(f64),
+}
+
+impl Display for PreparedMetric {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PreparedMetric::Unsigned(e) => write!(f, "{}", e),
+            PreparedMetric::Float(e) => write!(f, "{}", e),
+        }
+    }
+}
+
+pub struct TagObserver;
+
+impl TagObserver {
+    pub async fn log_tag_metrics(log_state: LogWeak, recorder: Arc<MmRecorder>, interval: f64) {
+        loop {
+            Timer::sleep(interval).await;
+            let log_state = match LogArc::from_weak(&log_state) {
+                Some(la) => la,
+                _ => {
+                    return;
+                },
+            };
+
+            log!(">>>>>>>>>> DEX metrics <<<<<<<<<");
+
+            Self::prepare_tag_metrics(&recorder)
+                .into_iter()
+                .for_each(|(labels, name_value_map)| {
+                    let tags = labels_to_tags(labels.iter());
+                    let message = name_value_map_to_message(&name_value_map);
+                    log_state.log_deref_tags("", tags, &message);
+                });
+
+            let quantiles = parse_quantiles(QUANTILES);
+            Self::prepare_tag_histograms(&recorder).into_iter().for_each(|histo| {
+                histo.into_iter().for_each(|(key, hist)| {
+                    let tags = labels_to_tags(Key::from_name(KeyName::from(key.clone())).labels());
+                    let message = format!("{}: {}", key, hist_to_message(quantiles.as_slice(), hist));
+                    log_state.log_deref_tags("", tags, &message);
+                });
+            });
```

Comment:
First of all, I've checked if we still can use the `hdrhistogram` crate. It takes metrics as `&[u64]`, but we expect to work with it with `&[f64]`.
Please read more about [histograms](https://github.com/HdrHistogram/HdrHistogram_rust).
You can notice that we use only two quantiles:
```rust
const QUANTILES: &[f64] = &[0.0, 1.0];
```
It's because we want to get MIN and MAX values among all timing metrics given by the same labels and metric value.
In the latest versions, `metrics_util` provides its own `Histogram`, but it seems to be inefficient if we're going to get MIN and MAX values only.
So I'd like to suggest adding our own `Histogram`:
```rust
struct Histogram {
    count: usize,
    min: f64,
    max: f64,
}

impl Histogram {
    fn new(data: &[f64]) -> Option<Histogram>;

    fn to_tag_message(&self) -> String { format!("count={} min={} max={}", self.count, self.min, self.max) }

    /// This method will be used in order to construct `MetricType::Histogram::quantiles`.
    // @borngraced please note that `MetricType::Histogram::quantiles` had `HashMap<String, u64>` type before.
    // Now it should be `HashMap<String, f64>`.
    fn to_json_quantiles(&self) -> HashMap<String, f64> {
        let mut result = HashMap::new();
        result.insert("count".to_owned(), self.count as f64);
        result.insert("min".to_owned(), self.min);
        result.insert("max".to_owned(), self.max);
    }
}
```

Then you can add new type variant `PreparedMetric::Histogram(Histogram)`. Please note that I guess there is no need to implement `fmt::Display` for `PreparedMetric`.

And then you need to remove `TagObserver::prepare_tag_histograms` and prepare the histograms within `TagObserver::prepare_tag_metrics`:
```rust
fn prepare_tag_metrics(recorder: &MmRecorder) -> HashMap<MetricLabels, MetricNameValueMap> {
        let mut output = HashMap::new();

        for (key, counter) in recorder.registry.get_counter_handles() {
            let value = counter.get_inner().load(Ordering::Acquire);
            map_metrics_to_prepare_tag_metric_output(key, PreparedMetric::Unsigned(value), &mut output);
        }

        for (key, gauge) in recorder.registry.get_gauge_handles() {
            let value = f64::from_bits(gauge.get_inner().load(Ordering::Acquire));
            map_metrics_to_prepare_tag_metric_output(key, PreparedMetric::Float(value), &mut output);
        }

        for (key, histogram) in recorder.registry.get_histogram_handles() {
            if let Some(hist) = Histogram::new(&histogram.get_inner().data()) {
                map_metrics_to_prepare_tag_metric_output(key, PreparedMetric::Histogram(hist), &mut output);
            }
        }

        output
    }
```